File: system/src/webql/engine/basic_ffi.erl
-module(basic_ffi).

-export([handle_run/1, handle_start_plan/1, handle_finish_plan/2, handle_start_batch/1,
         handle_finish_batch/3, handle_start_step/1, handle_finish_step/2]).

handle_run(Next) ->
  case Next() of
    {ok, Task} ->
      await(Task);
    Error ->
      Error
  end.

handle_start_plan(Next) ->
  async(fun() ->
           case Next() of
             {ok, {Initial, Batches}} ->
               lists:foldl(fun (Batch, {ok, Memory}) -> await(Batch(Memory));
                               (_Batch, Error) -> Error
                           end,
                           {ok, Initial},
                           Batches);
             Error -> Error
           end
        end).

handle_finish_plan(Task, Next) ->
  async(fun() ->
           case await(Task) of
             {ok, Memory} -> Next(Memory);
             Error -> Error
           end
        end).

handle_start_batch(Next) ->
  Next().

handle_finish_batch(Initial, Task, Merge) ->
  async(fun() ->
           case await(Task) of
             {ok, Steps} ->
               lists:foldl(fun ({ok, StepMemory}, {ok, Memory}) -> {ok, Merge(Memory, StepMemory)};
                               (Error, {ok, _Memory}) -> Error;
                               (_Step, Error) -> Error
                           end,
                           {ok, Initial},
                           lists:map(fun await/1, Steps));
             Error -> Error
           end
        end).

handle_start_step(Next) ->
  async(fun() ->
           case Next() of
             {ok, Task} -> await(Task);
             Error -> Error
           end
        end).

handle_finish_step(Task, Next) ->
  async(fun() -> Next(await(Task)) end).

%% PRIVATE FUNCTIONS
%% =================
async(Fun) ->
  {async,
   spawn(fun() ->
            Result =
              try
                {ok, Fun()}
              catch
                Kind:Reason:Stacktrace -> {error, Kind, Reason, Stacktrace}
              end,
            pending(Result)
         end)}.

await({async, Pid}) ->
  Ref = make_ref(),
  Monitor = erlang:monitor(process, Pid),
  Pid ! {self(), Ref},
  receive
    {Ref, {ok, Result}} ->
      erlang:demonitor(Monitor, [flush]),
      Result;
    {Ref, {error, Kind, Reason, Stacktrace}} ->
      erlang:demonitor(Monitor, [flush]),
      erlang:raise(Kind, Reason, Stacktrace);
    {'DOWN', Monitor, process, Pid, Reason} ->
      exit(Reason)
  end;
await(Task) ->
  Task.

pending(Result) ->
  receive
    {Caller, Ref} ->
      Caller ! {Ref, Result},
      pending(Result)
  end.

File: webql/src/webql/engine/basic_ffi.mjs
import { Empty, Ok, Error } from "../../gleam.mjs";

// NOTE: Gleam impliments a number-based access for their custom classes.
const ACCESS_INDEX = 0;

export function handleRun(next) {
  const result = next();
  return result instanceof Ok ? result[ACCESS_INDEX] : result;
}

export async function handleStartPlan(next) {
  const result = next();

  if (result instanceof Error) {
    return result;
  }

  const [initial, pendingBatches] = result[ACCESS_INDEX];
  const batches =
    pendingBatches instanceof Empty
      ? []
      : [pendingBatches.head, ...toArray(pendingBatches.tail)];

  return batches.reduce(
    async (acc, batch) => {
      const resolved = await acc;
      return resolved instanceof Error
        ? resolved
        : batch(resolved[ACCESS_INDEX]);
    },
    Promise.resolve(new Ok(initial)),
  );
}

export async function handleFinishPlan(task, next) {
  const result = await task;
  return result instanceof Ok ? next(result[ACCESS_INDEX]) : result;
}

export function handleStartBatch(next) {
  return next();
}

export async function handleFinishBatch(initial, task, merge) {
  const result = await task;

  if (result instanceof Error) {
    return result;
  }

  const steps = await Promise.all(result[ACCESS_INDEX]);

  return steps.reduce((acc, step) => {
    if (acc instanceof Error) {
      return acc;
    }

    if (step instanceof Error) {
      return step;
    }

    return new Ok(merge(acc[ACCESS_INDEX], step[ACCESS_INDEX]));
  }, new Ok(initial));
}

export function handleStartStep(next) {
  const result = next();
  return result instanceof Ok ? result[ACCESS_INDEX] : result;
}

export async function handleFinishStep(task, next) {
  return next(await task);
}

const toArray = (list) =>
  list instanceof Empty ? [] : [list.head, ...toArray(list.tail)];
